Engine/Helper.cs

```diff
@@ -444,41 +458,54 @@ public string GetTypeFromReturnStatementAst(Ast funcAst, ReturnStatementAst ret,
                 {
                     if (cmAst.Expression.StaticType != typeof(object))
                     {
-                        return cmAst.Expression.StaticType.FullName;
+                        result = cmAst.Expression.StaticType.FullName;
                     }
-
-                    VariableExpressionAst varAst = cmAst.Expression as VariableExpressionAst;
-
-                    if (varAst != null)
+                    else
                     {
-                        return GetVariableTypeFromAnalysis(varAst, funcAst);
+                        VariableExpressionAst varAst = cmAst.Expression as VariableExpressionAst;
+
+                        if (varAst != null)
+                        {
+                            result = GetVariableTypeFromAnalysis(varAst, funcAst);
+                        }
+                        else if (cmAst.Expression is MemberExpressionAst)
+                        {
+                            result = GetTypeFromMemberExpressionAstHelper(cmAst.Expression as MemberExpressionAst, funcAst, classes);
+                        }
                     }
+                }
+            }
 
-                    MemberExpressionAst memAst = cmAst.Expression as MemberExpressionAst;
+            if (String.IsNullOrWhiteSpace(result) && pipe != null && pipe.PipelineElements.Count > 0)
+            {
+                result = typeof(object).FullName;
+            }
 
-                    if (memAst != null)
-                    {
-                        VariableAnalysisDetails details = null;
-                        TypeDefinitionAst psClass = null;
+            return result;
+        }
 
-                        if (memAst.Expression is VariableExpressionAst && VariableAnalysisDictionary.ContainsKey(scriptAst))
-                        {
-                            VariableAnalysis VarTypeAnalysis = VariableAnalysisDictionary[scriptAst];
-                            details = VarTypeAnalysis.GetVariableAnalysis(memAst.Expression as VariableExpressionAst);
+        internal string GetTypeFromMemberExpressionAstHelper(MemberExpressionAst memberAst, Ast scopeAst, IEnumerable<TypeDefinitionAst> classes)
```

Comment:
I think it looks fine here but if we want to add the component back, you may need to add comments for this method.